xarray/core/computation.py

```diff
@@ -0,0 +1,261 @@
+    if is_scalar(arg):
+        return cls([arg])
+    else:
+        return cls(arg)
+
+
+_ElemwiseSignature = NamedTuple(
+    '_ElemwiseSignature', 'broadcast_dims, core_dims, output_dims, axis')
+
+
+def _build_and_check_signature(variables, core_dims=None, axis_dims=None,
+                               drop_dims=None, new_dims=None):
+    # All input dimension arguments are checked to appear on at least one input:
+    # - core_dims are not broadcast over, and moved to the right with order
+    #   preserved.
+    # - axis_dims is used to generate an integer or tuples of integers `axis`
+    #   keyword argument, which corresponds to the position of the given
+    #   dimension on the inputs. If `axis_dims` have overlap with `core_dims`,
+    #   no non-axis dimensions may appear in `core_dims` before an axis
+    #   dimension.
+    # - drop_dims are input dimensions that should be dropped from the output.
+    #
+    # All output dimensions arguments are checked not to appear on any inputs:
+    # - new_dims are new dimensions that should be added to the output array, in
+    #   order to the right of dimensions that are not dropped.
+
+    if core_dims is None and drop_dims is None and axis_dims is None:
+        # broadcast everything
+        dims = tuple(_calculate_unified_dim_sizes(variables))
+        return _ElemwiseSignature(dims, (), dims, None)
+
+    core_dims = () if core_dims is None else _as_sequence(core_dims, tuple)
+    drop_dims = set() if drop_dims is None else _as_sequence(drop_dims, set)
+    new_dims = () if new_dims is None else _as_sequence(new_dims, tuple)
+
+    axis_is_scalar = axis_dims is not None and is_scalar(axis_dims)
+    axis_dims = set() if axis_dims is None else _as_sequence(axis_dims, set)
+
+    dim_sizes = _calculate_unified_dim_sizes(variables)
+    broadcast_dims = tuple(d for d in dim_sizes if d not in core_dims)
+    all_input_dims = set(dim_sizes)
+
+    invalid = set(core_dims) - all_input_dims
+    if invalid:
+        raise ValueError('some `core_dims` not found on any input variables: '
+                         '%r' % list(invalid))
+
+    invalid = drop_dims - all_input_dims
+    if invalid:
+        raise ValueError('some `drop_dims` not found on any input variables: '
+                         '%r' % list(invalid))
+
+    invalid = axis_dims - all_input_dims
+    if invalid:
+        raise ValueError('some `axis_dims` not found on any input variables: '
+                         '%r' % list(invalid))
+    axis = tuple(broadcast_dims.index(d) for d in axis_dims)
+    n_remaining_axes = len(axis_dims) - len(axis)
+    if n_remaining_axes > 0:
+        valid_core_dims_for_axis = core_dims[:remaining_axes]
+        if not set(valid_core_dims_for_axis) <= axis_dims:
+            raise ValueError('axis dimensions %r have overlap with core '
+                             'dimensions %r, but do not appear at the start'
+                             % (axis_dims, core_dims))
```

Comment:
Why does the order of the passed dims matter? (i.e. why not transpose them all into the order that's needed)